### ISSUE TITLE
Is there a way to store Doorkeeper::AccessToken data inside Doorkeeper::JWT (and not use db-backed tokens)?

### ISSUE THREAD

**Issue by NeilSlater-Clixifix** [doorkeeper-gem/doorkeeper#1600]
### Steps to reproduce

Not a bug report, but a feature request for relying on token data storage *inside* issued JWT

I *think* this is either an undocumented existing feature (in which case I would appreciate a "how to" link), or more likely something that Doorkeeper 5.5 does not support.

### Expected behavior

There should be a way to inform Doorkeeper / Doorkeeper::JWT to not use the`Doorkeeper::AccessToken` database model, and that instead the details such as client id, approved scopes etc should be stored and read direct from the JWT. This is a common use-case for JWTs as an authentication mechanism, that they *embed* data for access control.

### Actual behavior

As far as I can tell from a quick browse of the code and local testing, Doorkeeper always uses database-backed access tokens, and JWTs are not exceptions - the whole serialised JWT is used as an index string to lookup server-side credentials. That means for busy API services, we cannot explore the possibility of skipping the db I/O for granting and reading access tokens.

### System configuration

**Doorkeeper initializer**:

```ruby
# config/initializers/doorkeeper.rb
Doorkeeper.configure do
  access_token_generator '::Doorkeeper::JWT'
end
```

**Doorkeeper JWT initializer**:

```ruby
# config/initializers/doorkeeper_jwt.rb

Doorkeeper::JWT.configure do
  token_payload do |opts|
    {
      iss: SITE,
      iat: Time.now.utc.to_i,

      # see JWT reserved claims - https://tools.ietf.org/html/draft-jones-json-web-token-07#page-7
      jti: SecureRandom.uuid,

      # Illustrative example (client credentials in this case, so no User information)
      doorkeeper_access: {
        application_id: 2,
        expires_in: 7200,
        scopes: "core_api read write"
      }
    }
  end
end
```

User and Doorkeeper::Application data could be similarly embedded on granting the token. This will make the token longer, but allow clients to skip more database reads on auth checks. Usually this would be combined with shorter token expiry times, and matches patterns of use for client apps that repeatedly make "one shot" requests and don't store access tokens.

**Comment by nbulaj** [comment]
Currently there is now way to use just raw JWT tokens without database (and it's not possible for Doorkeeper by design as we can't revoke such tokens and do some other things which is a core of the gem)

**Comment by nbulaj** [comment]
If anybody interested in this feature - I can suggest to create some `doorkeeper-no-orm` extension which will implement required ORM interface (at least what is possible) and use JWT tokens instead of database storage. Examples could be found in doorkeeper-sequel / doorkeeper-mongodb / etc projects

**Comment by nbulaj** [comment]
Hey @mschoenlaub . I'm not sure how exactly you plan to extend `doorkeeper-jwt` gem. Maybe only if you'll introduce a `JWT` ORM inside it :smiley_cat: Because all the interactions with the tokens are performed using the models (like find, create, revoke and so on). Some of the features will be missed for sure (`reuse_access_token` , token revocation and many more).

You can try to if you want, I'll be happy to review it :bow:

**Comment by 55728** [comment]
Keeping this open 👀✨ — there's room for future extension here. Removed the spec and `Closes` reference from #1901.

**Comment by nbulaj** [comment]
Just updated with `main` and pushed my branch with a stateless token - https://github.com/doorkeeper-gem/doorkeeper/tree/stateless-jwt-token

I never tested it and TBH not sure if we wanna to evaluate and support it. But anyone who searches for it can at least check the idea

**Comment by 55728** [comment]
Hey @nbulaj — I checked out the `stateless-jwt-token` branch 👀✨

Really like the approach: the opaque/JWT coexistence via the 3-segment check is clean, and `StatelessToken` duck-typing the existing helpers means zero changes on the consumer side.

I added spec coverage for the uncovered paths and opened a PR against the branch: #1943

Happy to keep poking at this if you want to take it further 😊

**Comment by neilslater** [comment]
OP here. Just to add that I would still be interested to see this land.

My employer is using Doorkeeper and expanding its use across more API endpoints, and anything we can do to reduce db lookups for performance is a significant boost.

**Comment by NeilSlater-Clixifix** [comment]
OP here. Just to add that I would still be interested to see this land.

My employer is using Doorkeeper and expanding its use across more API endpoints, and anything we can do to reduce db lookups for performance is a significant boost.

**Comment by 55728** [comment]
Hi @nbulaj 👋

Following up on my earlier comment — I ran end-to-end request specs against the `stateless-jwt-token` branch (`8483ffb8`) to evaluate it further. The unit specs (126 examples, including the ones I added in #1943) all pass, but the flow-level testing turned up several issues.

You mentioned you never tested it and weren't sure whether you'd want to support it — so treating this as input for that decision rather than a bug report 👀✨

## TL;DR

**The `authorization_code` grant raises `NoMethodError` (500) on a fresh install with the default migration.** Beyond that, there are gaps around revocation, introspection, and audience restriction.

## Findings summary

| # | Severity | Issue |
|---|----------|-------|
| 1 | **Critical** | `authorization_code` exchange → 500. `link_access_token_to_grant` calls `access_token.id`, but `StatelessToken` has no `#id`. The guard checks `access_token_revoked_on_reuse?` (= whether the `access_token_id` column exists), and the default migration includes it, so every fresh install hits this. |
| 2 | High | `/oauth/revoke` returns 200 and sets `revoked_at` on the DB row, but the token keeps working because the stateless auth path never checks the DB. Affects any token that has a DB row — tokens issued before stateless mode was enabled, and `client_credentials` tokens (see item 3). |
| 3 | High | `client_credentials` bypasses stateless entirely. `Creator#call` calls `create_for` directly instead of `find_or_create_for`, so it always writes to DB. |
| 4 | High | RFC 8707 `resource` indicator is silently dropped — never written to claims or passed to the generator. Audience-restricted grants produce unrestricted tokens. |
| 5 | Medium | Generator receives fewer attributes on the stateless path than the AR path (`resource_owner` for polymorphic, custom attributes are missing). |
| 6 | Medium | Introspection returns `{"active": false}` for valid stateless tokens (looks up DB only). |
| 7 | Medium | A JWT without `exp` is accepted and lives forever with no way to invalidate it. |
| 8 | Medium | `claims["sub"]` fallback + no `typ`/`aud` verification means a JWT signed with the same key (e.g. an `id_token`) could pass as an access token on endpoints using `doorkeeper_authorize!` without scope restriction. |

Findings 1, 2, 3, 5 and 6 are reproduced end-to-end with a scratch request spec (HS256 generator + `JWT.decode` decoder); 4, 7 and 8 are from reading the code path.

Additionally:
- No CHANGELOG / README / UPGRADE entry, and the expected claim names for `jwt_token_decoder` (`resource_owner_id` / `scope` / `client_id` / `exp`) aren't documented anywhere. `doorkeeper-jwt`'s default `token_payload` doesn't produce these, so there's a gap between the documented setup and what the stateless path actually expects.
- No flow-level request specs — `spec/requests/flows/` has no stateless coverage, which is why findings 1 and 3 weren't caught at the unit level.
- The branch is based on `29db4223` and is behind `upstream/main` (`5dbac564`).
- Config validation for `jwt_token_decoder` only warns instead of raising `Errors::InvalidConfiguration` like other validations do, so a missing decoder silently makes every Bearer request 401.

## Design decisions

**0. Do we want to support this at all?** If the answer is no, the rest is moot and I'd suggest documenting the branch as an experiment in the issue and closing it out. The findings below are what it would cost to make it supportable.

If yes:

1. **`authorization_code` + stateless + RFC 6749 §10.5**: The code-reuse revocation defense (`link_access_token_to_grant`) can't work without a persisted token. Options: (a) skip the link and accept that the defense is grant-revocation only, or (b) reject `authorization_code` + stateless at boot time.

2. **`/oauth/revoke` and `/oauth/introspect` behavior**: (a) return `unsupported_token_type` for stateless tokens (my preference — stays consistent with the "no DB" philosophy), or (b) add a DB fallback for revocation checks.

3. **Claim name contract**: The stateless path hard-codes `resource_owner_id` / `scope` / `client_id` / `exp`, but these aren't documented and don't match `doorkeeper-jwt`'s default `token_payload`. Should we formalize our own contract, or align with RFC 9068 (`at+jwt`)?

## What I can do

For finding 1 specifically, I can send a minimal guard (`return unless access_token.respond_to?(:id)`) right away so the branch isn't broken while we discuss Q1 — it stops the 500 without deciding the §10.5 question, and I'd note the trade-off in the PR body.

Beyond that, once we agree on direction I'm happy to put together PRs — probably 5–6 covering the fixes, flow specs, and documentation. I'd start with the straightforward bug fixes (findings 3, 4, 5) that don't depend on the design questions, then tackle the rest.

Full reproduction steps and proposed diffs are in my local report if you'd like me to share any of them.

@NeilSlater-Clixifix — appreciate the nudge 🙂 There's real work left, but the foundation is solid and I'm keen to help get it into shape. Worth flagging for your use case: finding 3 means `client_credentials` — probably the most read-heavy path for a machine-to-machine API — still hits the DB on every issuance today, so the performance win isn't there yet.
